src/graphic/Pattern.ts

```diff
@@ -5,16 +5,16 @@ type CanvasPatternRepeat = 'repeat' | 'repeat-x' | 'repeat-y' | 'no-repeat'
 export interface PatternObject {
     id?: number
 
-    type: 'pattern'
+    type?: 'pattern'
 
     image: ImageLike | string
```

Comment:
Is it better to use `union of interface`?
```ts
interface CanvasPatternObject {
  type: 'pattern'
  image: ImageLike | string
  repeat: ...
}
interface SVGPattenObject  {
  type: 'pattern'
  svgElement?: SVGElement
  ....
}

type PatternObject = CanvasPatternObject | SVGPatternObject
```

Reply:
I'm not sure which way is better,

1) Use union types (you mentioned above)
2) Define a parent type named `PatternObject`, `CanvasPatternObject ` and `SVGPatternObject` extend it.

Reply:
Union types provide a more precise type checking for the developers. It's usually better when the subtypes are enumerable.

For `Gradient` I now prefer the interface to be:

```ts
export interface GradientObjectBase {
    id?: number
    type: string
    colorStops: GradientColorStop[]
    global: boolean
}

export interface LinearGradientObject extends GradientObjectBase { };
export interface RadiaGradientObject extends GradientObjectBase { };
export type GradientObject = LinearGradientObject | RadiaGradientObject;
```